### PR TITLE
[interpreter] Implement bulk memory operations

### PR DESCRIPTION
**Pull request by binji**
This adds the new operations, as well as the new data and element
segment formats.

Implemented:

* Text parsing/writing
* Binary encoding/decoding
* Validation

TODO:

* Tests
* DataCount section
* Proper encoding of passive element segments (using ref.null etc)
* Evaluation

### CODE REVIEW

File: interpreter/text/parser.mly [hunk 1]
```diff
@@ -730,13 +772,13 @@ module_fields1 :
       { m with funcs = funcs @ m.funcs;
         imports = ims @ m.imports; exports = exs @ m.exports } }
   | elem module_fields
-    { fun c -> let mf = $2 c in
-      fun () -> let m = mf () in
-      {m with elems = $1 c :: m.elems} }
+    { fun c -> let ef = $1 c in let mf = $2 c in
```
Comment (rossberg) [inline interpreter/text/parser.mly]: Guessing you had a lot of fun figuring out the right phasing here. :) Sorry for the mess.

File: interpreter/binary/decode.ml [hunk 2]
```diff
@@ -589,10 +607,24 @@ let code_section s =
 (* Element section *)
 
 let segment dat s =
-  let index = at var s in
```
[the remark this answers is not in the document]
Comment (binji) [reply interpreter/binary/decode.ml]: In the overview this is a `vu32`, fixed it here so it's not a breaking change.

File: interpreter/valid/valid.ml [hunk 3]
```diff
@@ -29,6 +27,7 @@ type context =
   results : value_type list;
   labels : stack_type list;
 }
+and segment_type = Seg
```
Comment (rossberg) [inline interpreter/valid/valid.ml]: Move this to types.ml?

File: interpreter/syntax/ast.ml [hunk 4]
```diff
@@ -147,7 +147,7 @@ and memory' =
 type 'data segment = 'data segment' Source.phrase
 and 'data segment' =
 {
-  stype : segment_type;
+  desc : segment_desc;
```
Comment (rossberg) [inline interpreter/syntax/ast.ml]: Since there are other xdesc's in the AST already, I'd call this `sdesc`.

**Pull request merged**